Code/Engine/Core/System/Implementation/glfw/ControllerInput_glfw.inl

```diff
@@ -0,0 +1,227 @@
+
+  virtual void UpdateInputSlotValues() override;
+
+  virtual void ResetInputSlotValues() override;
+
+  virtual void RegisterInputSlots() override;
+
+  virtual bool IsControllerConnected(ezUInt8 uiPhysical) const override;
+
+private:
+  virtual void ApplyVibration(ezUInt8 uiPhysicalController, Motor::Enum eMotor, float fStrength) override;
+
+  void RegisterControllerButton(const char* szButton, const char* szName, ezBitflags<ezInputSlotFlags> SlotFlags);
+  void SetDeadZone(const char* szButton);
+  void SetControllerValue(ezStringBuilder& tmp, ezUInt8 controllerIndex, const char* inputSlotName, float value);
+};
+
+namespace
+{
+  ezUniquePtr<ezControllerInputGlfw> g_pControllerInputGlfw;
+}
+
+// clang-format off
+EZ_BEGIN_SUBSYSTEM_DECLARATION(Core, ControllerInput)
+
+  BEGIN_SUBSYSTEM_DEPENDENCIES
+    "Foundation",
+    "InputManager",
+    "Window"
+  END_SUBSYSTEM_DEPENDENCIES
+
+  ON_CORESYSTEMS_STARTUP
+  {
+    g_pControllerInputGlfw = EZ_DEFAULT_NEW(ezControllerInputGlfw);
+    ezControllerInput::SetDevice(g_pControllerInputGlfw.Borrow());
+  }
+
+  ON_CORESYSTEMS_SHUTDOWN
+  {
+    if (ezControllerInput::GetDevice() == g_pControllerInputGlfw.Borrow())
+    {
+      ezControllerInput::SetDevice(nullptr);
+    }
+    g_pControllerInputGlfw.Clear();
+  }
+
+  ON_HIGHLEVELSYSTEMS_STARTUP
+  {
+  }
+
+  ON_HIGHLEVELSYSTEMS_SHUTDOWN
+  {
+  }
+
+EZ_END_SUBSYSTEM_DECLARATION;
+// clang-format on
+
+namespace
+{
+  struct ControllerButtonMapping
+  {
+    const char* ezName;
+    int glfwIndex;
+  };
+
+  const ControllerButtonMapping g_ControllerButtonMappings[] = {
+    {"button_a", GLFW_GAMEPAD_BUTTON_A},
+    {"button_b", GLFW_GAMEPAD_BUTTON_B},
+    {"button_x", GLFW_GAMEPAD_BUTTON_X},
+    {"button_y", GLFW_GAMEPAD_BUTTON_Y},
+    {"button_start", GLFW_GAMEPAD_BUTTON_START},
+    {"button_back", GLFW_GAMEPAD_BUTTON_BACK},
+    {"left_shoulder", GLFW_GAMEPAD_BUTTON_LEFT_BUMPER},
+    {"right_shoulder", GLFW_GAMEPAD_BUTTON_RIGHT_BUMPER},
+    {"pad_up", GLFW_GAMEPAD_BUTTON_DPAD_UP},
+    {"pad_down", GLFW_GAMEPAD_BUTTON_DPAD_DOWN},
+    {"pad_left", GLFW_GAMEPAD_BUTTON_DPAD_LEFT},
+    {"pad_right", GLFW_GAMEPAD_BUTTON_DPAD_RIGHT},
+    {"left_stick", GLFW_GAMEPAD_BUTTON_LEFT_THUMB},
+    {"right_stick", GLFW_GAMEPAD_BUTTON_RIGHT_THUMB}};
+} // namespace
+
+void ezControllerInputGlfw::RegisterControllerButton(const char* szButton, const char* szName, ezBitflags<ezInputSlotFlags> SlotFlags)
+{
+  ezStringBuilder s, s2;
+
+  for (ezInt32 i = 0; i < MaxControllers; ++i)
+  {
+    s.Format("controller{0}_{1}", i, szButton);
+    s2.Format("Cont {0}: {1}", i + 1, szName);
+    RegisterInputSlot(s.GetData(), s2.GetData(), SlotFlags);
+  }
+}
+
+void ezControllerInputGlfw::SetDeadZone(const char* szButton)
+{
+  ezStringBuilder s;
+
+  for (ezInt32 i = 0; i < MaxControllers; ++i)
+  {
+    s.Format("controller{0}_{1}", i, szButton);
+    ezInputManager::SetInputSlotDeadZone(s.GetData(), 0.23f);
```

Comment:
what's the purpose of this ?

Reply:
I copied the SetDeadZone code straight from the XBox controller plugin. I though this is "the way to go"